src/Tools/dotnet-monitor/CollectionRules/Actions/CollectDumpAction.cs

```diff
@@ -3,19 +3,80 @@
 // See the LICENSE file in the project root for more information.
 
 using Microsoft.Diagnostics.Monitoring.WebApi;
+using Microsoft.Diagnostics.Monitoring.WebApi.Controllers;
+using Microsoft.Diagnostics.Monitoring.WebApi.Models;
+using Microsoft.Diagnostics.Tools.Monitor.CollectionRules.Exceptions;
 using Microsoft.Diagnostics.Tools.Monitor.CollectionRules.Options.Actions;
+using Microsoft.Extensions.DependencyInjection;
 using System;
+using System.Collections.Generic;
 using System.Threading;
 using System.Threading.Tasks;
 
 namespace Microsoft.Diagnostics.Tools.Monitor.CollectionRules.Actions
 {
-    internal sealed class CollectDumpAction :
-        ICollectionRuleAction<CollectDumpOptions>
+    internal sealed class CollectDumpAction : ICollectionRuleAction<CollectDumpOptions>
     {
-        public Task<CollectionRuleActionResult> ExecuteAsync(CollectDumpOptions options, IEndpointInfo endpointInfo, CancellationToken token)
+        private readonly IDiagnosticServices _diagnosticServices;
+        private readonly IServiceProvider _serviceProvider;
+
+        public CollectDumpAction(IServiceProvider serviceProvider)
+        {
+            _serviceProvider = serviceProvider;
+            _diagnosticServices = serviceProvider.GetRequiredService<IDiagnosticServices>();
+        }
+
+        public async Task<CollectionRuleActionResult> ExecuteAsync(CollectDumpOptions options, IEndpointInfo endpointInfo, CancellationToken token)
+        {
+            DumpType dumpType = options.Type.GetValueOrDefault(CollectDumpOptionsDefaults.Type);
+            string egressProvider = options.Egress;
+
+            IProcessInfo processInfo = await _diagnosticServices.GetProcessAsync(new ProcessKey(endpointInfo.ProcessId), token);
+
+            string dumpFileName = DiagController.GenerateDumpFileName();
+
+            string dumpFilePath = "";
+
+            // Given our options validation, I believe this is probably redundant...should I remove it?
+            if (string.IsNullOrEmpty(egressProvider))
+            {
+                // Also, I would move this to Strings.resx if we do keep it, but I decided to wait for feedback before doing that.
+                throw new ArgumentException("No Egress Provider was supplied.");
+            }
+            else
+            {
+                KeyValueLogScope scope = DiagController.GetDumpScope(processInfo);
+
+                try
+                {
+                    dumpFilePath = await SendToEgress(new EgressOperation(
```

Comment:
Since we are going to have to egress from many other actions, it would be very useful if we had an egress facade that could be used by every action.